http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/http2/Http2MultiplexedChannelPool.java

```diff
@@ -46,151 +59,335 @@
+                             .exceptionally(throwable -> failAndCloseParent(promise, parentChannel, throwable));
+            } catch (Throwable e) {
+                failAndCloseParent(promise, parentChannel, e);
+            }
+        });
+    }
+
+    private void acquireStreamOnFreshConnection(Promise<Channel> promise, Channel parentChannel, Protocol protocol) {
+        try {
+            Long maxStreams = parentChannel.attr(ChannelAttributeKey.MAX_CONCURRENT_STREAMS).get();
+
+            Validate.isTrue(protocol == Protocol.HTTP2,
+                            "Protocol negotiated on connection (%s) was expected to be HTTP/2, but it "
+                            + "was %s.", parentChannel, Protocol.HTTP1_1);
+            Validate.isTrue(maxStreams != null,
+                            "HTTP/2 was negotiated on the connection (%s), but the maximum number of "
+                            + "streams was not initialized.", parentChannel);
+            Validate.isTrue(maxStreams > 0, "Maximum streams were negative on channel (%s).", parentChannel);
+
+            MultiplexedChannelRecord multiplexedChannel = new MultiplexedChannelRecord(parentChannel, maxStreams);
+            parentChannel.attr(MULTIPLEXED_CHANNEL).set(multiplexedChannel);
+
+            Promise<Channel> streamPromise = parentChannel.eventLoop().newPromise();
+
+            if (!acquireStreamOnInitializedConnection(multiplexedChannel, streamPromise)) {
+                failAndCloseParent(promise, parentChannel,
+                                   new IOException("Connection was closed while creating a new stream."));
+                return;
+            }
+
+            streamPromise.addListener(f -> {
+                if (!streamPromise.isSuccess()) {
+                    promise.setFailure(streamPromise.cause());
+                    return;
+                }
+
+                Channel stream = streamPromise.getNow();
+                cacheConnectionForFutureStreams(stream, multiplexedChannel, promise);
+            });
+        } catch (Throwable e) {
+            failAndCloseParent(promise, parentChannel, e);
+        }
+    }
+
+    private void cacheConnectionForFutureStreams(Channel stream,
+                                                 MultiplexedChannelRecord multiplexedChannel,
+                                                 Promise<Channel> promise) {
+        Channel parentChannel = stream.parent();
+
+        // Before we cache the connection, make sure that exceptions on the connection will remove it from the cache.
+        parentChannel.pipeline().addLast(ReleaseOnExceptionHandler.INSTANCE);
+        connections.add(multiplexedChannel);
+
+        if (closed.get()) {
+            // Whoops, we were closed while we were setting up. Make sure everything here is cleaned up properly.
+            failAndCloseParent(promise, parentChannel,
+                               new IOException("Connection pool was closed while creating a new stream."));
```

Comment:
Can we do the removal from `connections` here?

Reply:
The failAndCloseParent will remove it. Is that what you mean?

Reply:
Yeah since we add it in here, I think it's this function's responsibility to remove it. I don't really feel strongly about it though